modules/table/src/test/java/org/apache/ignite/internal/table/distributed/IndexCleanupTest.java

```diff
@@ -0,0 +1,425 @@
+                new SortedIndexColumnDescriptor("INTVAL", NativeTypes.INT32, false, true),
+                new SortedIndexColumnDescriptor("STRVAL", NativeTypes.STRING, false, true)
+        )));
+
+        TableSchemaAwareIndexStorage sortedIndexStorage = new TableSchemaAwareIndexStorage(
+                sortedIndexId,
+                sortedInnerStorage,
+                USER_INDEX_BINARY_TUPLE_CONVERTER::toTuple
+        );
+
+        hashInnerStorage = new TestHashIndexStorage(new HashIndexDescriptor(hashIndexId, List.of(
+                new HashIndexColumnDescriptor("INTVAL", NativeTypes.INT32, false),
+                new HashIndexColumnDescriptor("STRVAL", NativeTypes.STRING, false)
+        )));
+
+        TableSchemaAwareIndexStorage hashIndexStorage = new TableSchemaAwareIndexStorage(
+                hashIndexId,
+                hashInnerStorage,
+                USER_INDEX_BINARY_TUPLE_CONVERTER::toTuple
+        );
+
+        storage = new TestMvPartitionStorage(1);
+
+        dataStorage = new TestPartitionDataStorage(storage);
+
+        storageUpdateHandler = new StorageUpdateHandler(1, dataStorage,
+                () -> Map.of(
+                        pkIndexId, pkStorage,
+                        sortedIndexId, sortedIndexStorage,
+                        hashIndexId, hashIndexStorage
+                )
+        );
+    }
+
+    @ParameterizedTest
+    @EnumSource(AddWrite.class)
+    void testAbort(AddWrite writer) {
+        UUID rowUuid = UUID.randomUUID();
+        RowId rowId = new RowId(1, rowUuid);
+
+        var key = new TestKey(1, "foo");
+        var value = new TestValue(2, "bar");
+        BinaryRow tableRow = binaryRow(key, value);
+
+        writer.addWrite(storageUpdateHandler, rowUuid, tableRow);
+
+        assertEquals(1, storage.rowsCount());
+        assertThat(pkInnerStorage.allRowsIds(), contains(rowId));
+        assertThat(sortedInnerStorage.allRowsIds(), contains(rowId));
+        assertThat(hashInnerStorage.allRowsIds(), contains(rowId));
+
+        storageUpdateHandler.handleTransactionAbortion(Set.of(rowId), () -> {});
+
+        assertEquals(0, storage.rowsCount());
+        assertTrue(pkInnerStorage.allRowsIds().isEmpty());
+        assertTrue(sortedInnerStorage.allRowsIds().isEmpty());
+        assertTrue(hashInnerStorage.allRowsIds().isEmpty());
+    }
+
+    @ParameterizedTest
+    @EnumSource(AddWrite.class)
+    void testTombstoneCleansUpIndexes(AddWrite writer) {
+        UUID rowUuid = UUID.randomUUID();
+
+        var key = new TestKey(1, "foo");
+        var value = new TestValue(2, "bar");
+        BinaryRow tableRow = binaryRow(key, value);
+
+        writer.addWrite(storageUpdateHandler, rowUuid, tableRow);
+        writer.addWrite(storageUpdateHandler, rowUuid, null);
+
+        // Write intent is in the storage.
+        assertEquals(1, storage.rowsCount());
+
+        // But indexes are removed.
```

Comment:
Shall we check that they were not empty after the first write?

Reply:
Right